### ISSUE TITLE
Inconsistent type inference when using `Any` and conditional expressions

### ISSUE THREAD

**Issue by RenDelaCruz** [python/mypy#16639]
**Bug Report**

There is inconsistency from the way mypy treats `Any` based on the order of a conditional expression. If `Any` is first, mypy ignores the rest of the conditional expression.

```py
condition = True
entry_name: Any = "entry_name"  # e.g. received from 3rd party library

name: str

# 1. This passes
name = entry_name if condition else None

# 2. And this passes too for some reason
name = 4 if condition else entry_name

# 3. But this is not allowed by mypy
name = None if condition else entry_name
```

But in reality, all 3 cases should not pass. **Or at least, for consistency, treat case (1) and (3) the same.** (ie. do not allow case 1)

**To Reproduce**

Mypy playground: https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=dcfde369054e28338aea63de87440fd2

**Expected Behavior**

- Should flag all 3 lines as not acceptable
- [Pyright example](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcAUCQMZgoAmSMSVUAvFACogCuApiVyjCDgB9FAEMIXAFyFizKACI%2BA4WInyoUAMRQuAOjS6oILuS5IAbl2pRQkKAGYQ1hKLxwoAGyQAjEK9IkqlJQAM4CZNoAjIasABZIIVAuISFcIYHiXHJKgiKZmMBQlDR0DCg6HqlQAHJUPCTaAEyGRNYw8YnJqYkwYGA2uKGQWcaiIVQZEnIALAVFVLT0jFyVWTkqmREOhgBCHPjtCZiJKGD4oh4eYADuVlDe7hCIpEFytShZSIXFi2UVVes8hISEA)

**Actual Behavior**

Mypy allows cases (1) and (2).

For case (3):

```sh
main.py:15: error: Incompatible types in assignment (expression has type "Any | None", variable has type "str")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.7.1
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.11

**Comment by ilevkivskyi** [comment]
Now mypy gives correct errors in all three examples.
